File: Services/PartnerSubscriptionSyncService.swift
import Foundation
import SwiftUI
import Combine
import FirebaseAuth
import FirebaseFirestore
import FirebaseFunctions

class PartnerSubscriptionSyncService: ObservableObject {
    static let shared = PartnerSubscriptionSyncService()
    
    private var cancellables = Set<AnyCancellable>()
    private var userListener: ListenerRegistration?
    private var partnerListener: ListenerRegistration?
    
    private init() {
        startListeningForUser()
        setupAuthObserver()
    }
    
    private func setupAuthObserver() {
        // Observer les changements d'authentification
        Auth.auth().addStateDidChangeListener { [weak self] (auth: Auth, user: FirebaseAuth.User?) in
            if user != nil {
                print("🔄 PartnerSubscriptionSyncService: Utilisateur reconnecté - Redémarrage listeners")
                DispatchQueue.main.asyncAfter(deadline: .now() + 1.0) {
                    self?.startListeningForUser()
                }
            } else {
                print("⚠️ PartnerSubscriptionSyncService: Utilisateur déconnecté - Arrêt listeners")
                self?.stopAllListeners()
            }
        }
    }
    
    // MARK: - Écoute des changements utilisateur
    
    private func startListeningForUser() {
        guard let currentUser = Auth.auth().currentUser else { 
            print("⚠️ PartnerSubscriptionSyncService: Utilisateur non authentifié, arrêt des listeners")
            return 
        }
        
        // Arrêter les anciens listeners
        stopAllListeners()
        
        userListener = Firestore.firestore()
            .collection("users")
            .document(currentUser.uid)
            .addSnapshotListener { [weak self] snapshot, error in
                if let error = error {
                    print("❌ PartnerSubscriptionSyncService: Erreur listener utilisateur: \(error)")
                    
                    // Si erreur de permissions, vérifier l'authentification
                    if error.localizedDescription.contains("permissions") {
                        print("⚠️ PartnerSubscriptionSyncService: Erreur de permissions - Vérification auth")
                        self?.handleAuthenticationError()
                    }
                    return
                }
                
                guard let data = snapshot?.data() else { 
                    print("⚠️ PartnerSubscriptionSyncService: Aucune donnée utilisateur trouvée")
                    return 
                }
                
                // 🔧 DEBUG: Logs détaillés pour identifier le problème
                let partnerIdRaw = data["partnerId"]
                print("🔍 PartnerSubscriptionSyncService: partnerId brut: \(partnerIdRaw ?? "nil")")
                print("🔍 PartnerSubscriptionSyncService: Type: \(type(of: partnerIdRaw))")
                
                // Si l'utilisateur a un partenaire connecté, écouter ses changements d'abonnement
                if let partnerId = data["partnerId"] as? String, !partnerId.isEmpty {
                    print("🔍 PartnerSubscriptionSyncService: partnerId valide trouvé: '\(partnerId)'")
                    self?.startListeningForPartner(partnerId: partnerId)
                } else {
                    print("🔍 PartnerSubscriptionSyncService: Aucun partenaire valide - arrêt écoute")
                    if let partnerIdString = data["partnerId"] as? String {
                        print("🔍 PartnerSubscriptionSyncService: partnerId est chaîne vide: '\(partnerIdString)'")
                    }
                    // Pas de partenaire, arrêter l'écoute
                    self?.stopListeningForPartner()
                }
            }
    }
    
    private func handleAuthenticationError() {
        print("🔧 PartnerSubscriptionSyncService: Gestion erreur authentification")
        
        // Arrêter tous les listeners
        stopAllListeners()
        
        // Vérifier si l'utilisateur est encore authentifié
        if Auth.auth().currentUser == nil {
            print("⚠️ PartnerSubscriptionSyncService: Utilisateur déconnecté - Arrêt du service")
        } else {
            print("🔄 PartnerSubscriptionSyncService: Redémarrage des listeners après erreur auth")
            // Redémarrer les listeners après un délai
            DispatchQueue.main.asyncAfter(deadline: .now() + 2.0) {
                self.startListeningForUser()
            }
        }
    }
    
    private func stopAllListeners() {
        userListener?.remove()
        userListener = nil
        stopListeningForPartner()
    }
    
    private func startListeningForPartner(partnerId: String) {
        // 🔧 CORRECTION: Vérification robuste du partnerId
        guard !partnerId.isEmpty, partnerId.trimmingCharacters(in: .whitespacesAndNewlines).count > 0 else {
            print("❌ PartnerSubscriptionSyncService: partnerId vide ou invalide reçu: '\(partnerId)'")
            stopListeningForPartner()
            return
        }
        
        // Arrêter l'ancienne écoute
        stopListeningForPartner()
        
        // 🔧 CORRECTION: Ne plus écouter directement les données du partenaire
        // La synchronisation se fera via Cloud Functions lors des changements d'abonnement
        print("🔄 PartnerSubscriptionSyncService: Synchronisation initiale avec partenaire: \(partnerId)")
        print("🔄 PartnerSubscriptionSyncService: partnerId validé - longueur: \(partnerId.count)")
        
        guard let currentUser = Auth.auth().currentUser else { return }
        
        // Faire une synchronisation initiale avec un petit délai pour éviter les conditions de course
        Task {
            // 🔧 CORRECTION: Petit délai pour laisser le temps aux mises à jour d'abonnement de se terminer
            try? await Task.sleep(nanoseconds: 2_000_000_000) // 2 secondes
            await self.syncSubscriptionViaCloudFunction(
                userId: currentUser.uid,
                partnerId: partnerId
            )
        }
    }
    
    private func stopListeningForPartner() {
        partnerListener?.remove()
        partnerListener = nil
    }
    
    // MARK: - Synchronisation des abonnements via Cloud Functions
    
    private func syncSubscriptionViaCloudFunction(userId: String, partnerId: String) async {
        // 🔧 CORRECTION: Vérification robuste du partnerId avant l'appel
        guard !partnerId.isEmpty, partnerId.trimmingCharacters(in: .whitespacesAndNewlines).count > 0 else {
            print("❌ PartnerSubscriptionSyncService: partnerId vide ou invalide: '\(partnerId)'")
            return
        }
        
        do {
            print("🔄 PartnerSubscriptionSyncService: Synchronisation via Cloud Function")
            print("🔄 PartnerSubscriptionSyncService: partnerId: '\(partnerId)' (longueur: \(partnerId.count))")
            
            let functions = Functions.functions()
            let data = [
                "partnerId": partnerId
            ]
            
            let result = try await functions.httpsCallable("syncPartnerSubscriptions").call(data)
            
            if let resultData = result.data as? [String: Any],
               let success = resultData["success"] as? Bool,
               success {
                print("✅ PartnerSubscriptionSyncService: Synchronisation réussie")
                
                // Notifier les changements si nécessaire
                if let inherited = resultData["subscriptionInherited"] as? Bool,
                   inherited,
                   let fromPartnerName = resultData["fromPartnerName"] as? String {
                    await MainActor.run {
                        NotificationCenter.default.post(
                            name: .partnerSubscriptionShared,
                            object: nil,
                            userInfo: [
                                "partnerId": userId,
                                "fromPartnerName": fromPartnerName
                            ]
                        )
                    }
                }
            } else {
                print("❌ PartnerSubscriptionSyncService: Échec synchronisation")
            }
            
        } catch {
            print("❌ PartnerSubscriptionSyncService: Erreur synchronisation Cloud Function: \(error)")
        }
    }
    

    
    deinit {
        userListener?.remove()
        partnerListener?.remove()
    }
    
    // MARK: - Méthode publique pour redémarrage
    
    func restart() {
        print("🔄 PartnerSubscriptionSyncService: Redémarrage manuel")
        stopAllListeners()
        startListeningForUser()
    }
} 

File: Services/PartnerSubscriptionSyncService.kt
package com.love2love.sync

import android.util.Log
import com.google.firebase.analytics.ktx.analytics
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.FirebaseUser
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.FirebaseFirestoreException
import com.google.firebase.firestore.ListenerRegistration
import com.google.firebase.functions.FirebaseFunctions
import com.google.firebase.ktx.Firebase
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.tasks.await

/**
 * Kotlin/Android port of PartnerSubscriptionSyncService (Swift).
 *
 * • Keeps a Firestore listener on current user doc to detect partnerId changes.
 * • When a valid partnerId appears, triggers a Cloud Function: "syncPartnerSubscriptions".
 * • Emits a local event when a subscription is inherited from the partner.
 * • Uses coroutines + Tasks.await() (add dependency: kotlinx-coroutines-play-services).
 * • Logs are "safe" (no raw UIDs printed).
 *
 * NOTE about localization: this service does not display UI strings.
 * For other parts that previously used `.xcstrings` like
 *   `challengeKey.localized(tableName: "DailyChallenges")`
 * on Android you must use `context.getString(R.string.challengeKey)`
 * or the Compose equivalent via `LocalContext.current`.
 */
object PartnerSubscriptionSyncService {

    private const val TAG = "PartnerSubscriptionSync"

    // Firebase
    private val auth: FirebaseAuth = FirebaseAuth.getInstance()
    private val db: FirebaseFirestore = FirebaseFirestore.getInstance()
    private val functions: FirebaseFunctions = FirebaseFunctions.getInstance()

    // Coroutines
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate)

    // Listeners
    private var authListener: FirebaseAuth.AuthStateListener? = null
    private var userListener: ListenerRegistration? = null
    private var partnerListener: ListenerRegistration? = null // kept for symmetry; not used (we sync via CF)

    // Event stream for UI to observe (subscription inherited, etc.)
    private val _subscriptionSharedEvents = MutableSharedFlow<PartnerSubscriptionSharedEvent>(extraBufferCapacity = 1)
    val subscriptionSharedEvents = _subscriptionSharedEvents.asSharedFlow()

    init {
        startAuthObserver()
    }

    // --- Public API

    fun restart() {
        Log.d(TAG, "Manual restart")
        stopAllListeners()
        startListeningForUser()
    }

    fun clear() {
        Log.d(TAG, "Clearing service")
        authListener?.let { auth.removeAuthStateListener(it) }
        authListener = null
        stopAllListeners()
    }

    // --- Internals

    private fun startAuthObserver() {
        if (authListener != null) return
        authListener = FirebaseAuth.AuthStateListener { firebaseAuth ->
            val user: FirebaseUser? = firebaseAuth.currentUser
            if (user != null) {
                Log.d(TAG, "User reconnected — restarting listeners")
                scope.launch {
                    delay(1_000)
                    startListeningForUser()
                }
            } else {
                Log.w(TAG, "User signed out — stopping listeners")
                stopAllListeners()
            }
        }
        auth.addAuthStateListener(authListener!!)
    }

    private fun startListeningForUser() {
        val currentUser = auth.currentUser ?: run {
            Log.w(TAG, "User not authenticated, stopping listeners")
            stopAllListeners()
            return
        }

        // Stop previous
        stopAllListeners()

        userListener = db.collection("users")
            .document(currentUser.uid)
            .addSnapshotListener { snapshot, error ->
                if (error != null) {
                    Log.e(TAG, "User listener error: ${error.message}", error)
                    if (error is FirebaseFirestoreException &&
                        error.code == FirebaseFirestoreException.Code.PERMISSION_DENIED) {
                        Log.w(TAG, "Permission error — checking auth")
                        handleAuthenticationError()
                    }
                    return@addSnapshotListener
                }

                val data = snapshot?.data
                if (data == null) {
                    Log.w(TAG, "No user data found")
                    return@addSnapshotListener
                }

                val partnerId = (data["partnerId"] as? String)?.trim().orEmpty()
                if (partnerId.isNotEmpty()) {
                    Log.d(TAG, "Valid partnerId found (length=${partnerId.length})")
                    startListeningForPartner(partnerId)
                } else {
                    Log.d(TAG, "No valid partner — stopping partner listener")
                    stopListeningForPartner()
                }
            }
    }

    private fun handleAuthenticationError() {
        Log.d(TAG, "Handling auth error")
        stopAllListeners()

        if (auth.currentUser == null) {
            Log.w(TAG, "User signed out — service idle")
        } else {
            Log.d(TAG, "Restarting listeners after auth error")
            scope.launch {
                delay(2_000)
                startListeningForUser()
            }
        }
    }

    private fun stopAllListeners() {
        userListener?.remove(); userListener = null
        stopListeningForPartner()
    }

    private fun startListeningForPartner(partnerId: String) {
        val pid = partnerId.trim()
        if (pid.isEmpty()) {
            Log.e(TAG, "Empty/invalid partnerId: '$partnerId'")
            stopListeningForPartner()
            return
        }

        // Stop previous partner listener if any (not used in this port)
        stopListeningForPartner()

        // Do NOT listen to partner doc directly — sync through Cloud Function
        Log.d(TAG, "Initial sync with partner (length=${pid.length})")

        val currentUser = auth.currentUser ?: return
        scope.launch(Dispatchers.IO) {
            // small delay to avoid race conditions with concurrent writes
            delay(2_000)
            syncSubscriptionViaCloudFunction(currentUser.uid, pid)
        }
    }

    private fun stopListeningForPartner() {
        partnerListener?.remove()
        partnerListener = null
    }

    private suspend fun syncSubscriptionViaCloudFunction(userId: String, partnerId: String) {
        val pid = partnerId.trim()
        if (pid.isEmpty()) {
            Log.e(TAG, "Empty/invalid partnerId: '$partnerId'")
            return
        }

        try {
            Log.d(TAG, "Sync via Cloud Function — with partner (length=${pid.length})")
            val data = hashMapOf(
                "partnerId" to pid
            )

            val result = functions
                .getHttpsCallable("syncPartnerSubscriptions")
                .call(data)
                .await()

            val map = result.data as? Map<*, *>
            val success = map?.get("success") as? Boolean ?: false
            if (success) {
                Log.d(TAG, "Sync success")
                val inherited = map["subscriptionInherited"] as? Boolean ?: false
                val fromPartnerName = map["fromPartnerName"] as? String

                if (inherited) {
                    // Analytics
                    Firebase.analytics.logEvent("abonnement_partage_partenaire", null)
                    Log.d(TAG, "Analytics event: abonnement_partage_partenaire")

                    // Emit local event to UI
                    _subscriptionSharedEvents.tryEmit(
                        PartnerSubscriptionSharedEvent(
                            partnerId = pid,
                            fromPartnerName = fromPartnerName ?: "Partenaire"
                        )
                    )
                }
            } else {
                Log.e(TAG, "Sync failed (success=false)")
            }
        } catch (t: Throwable) {
            Log.e(TAG, "Cloud Function sync error: ${t.message}", t)
        }
    }

    data class PartnerSubscriptionSharedEvent(
        val partnerId: String,
        val fromPartnerName: String
    )
}
